### PR TITLE
Feature/constraints model

### PR DESCRIPTION
**Pull request by gorold**
### Context
Completed models to handle price optimisation constraints and created basic tests.

### Changes
- Added Constraint model
- Added ConstraintParameter model
- Added ConstraintParameterRelationship model
- Removed Weight and RequiredHyperparameter models (currently not relevant, will add SavedModel model in future PR)
- Created basic tests

### Considerations
Will add SavedModel model and related endpoint in the future to handle trained models.
A ConstraintBlock is basically one set of many Constraints which will be used in a price optimisation process.
A Constraint is essentially a single equation, e.g. ```x + y <= 0```
A ConstraintBlock has several ConstraintParameters, which appear in the Constraint equations, thus, they have a ManyToMany relationship which is modeled using the ConstraintParameterRelationship.

### CODE REVIEW

File: secretsauce/apps/portal/models.py [hunk 1]
```diff
@@ -29,86 +29,84 @@ def __str__(self):
         return self.name
 
 class ModelTag(models.Model):
-    name = models.CharField(max_length=200)
+    name = models.CharField(max_length=200, unique=True, primary_key=True)
     prediction_model = models.ManyToManyField(
         PredictionModel,
-        related_name="tags"
+        related_name='model_tags',
     )
```
Comment (eugenechia95) [inline secretsauce/apps/portal/models.py:36]: What is ModelTag for again?

File: secretsauce/apps/portal/models.py [hunk 2]
```diff
@@ -29,86 +29,84 @@ def __str__(self):
...
     id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
-    data_block = models.ForeignKey(
-        DataBlock, 
-        on_delete=models.CASCADE, 
-        related_name="datablocks"
+    project = models.ForeignKey(
+        Project,
+        on_delete=models.CASCADE,
+        related_name='constraint_blocks',
     )
```
Comment (eugenechia95) [inline secretsauce/apps/portal/models.py:62]: Shouldn't constraint block be linked to datablock? Like different csv files will probably have different items so they should have different constraints too despite being in the same project.

**Comment by gorold** [comment]
> Just to clarify:
> 
> * So a ConstraintParameter's name is something like a particular item? By extension, all constraints in a constraint block will always have the same few constraint parameters to use in the equation?
> * What is the Constraint's name? Is it just a generic name for identification purposes e.g. Constraint1, Constraint2?
> * For the coefficient on ConstraintParameterRelationship, is it referring to "0" like in the example `x + y <= 0`?
> * If I understand correctly. This is the current class Diagram and my issue with it is there is too much interdependence between the classes:
>   image
> 
> Why not shift the coefficient as a field under Constraint instead and model it like that?
> image

- Yes, a ConstraintParameter will be like "Fish Burger Price", "Big Mac Price", etc. All Constraints in a ConstraintBlock should reuse ConstraintParameters.
- A Constraint's name can be like "Big Mac more expensive than fish burger", or "No arbitrage fries". I think Meng Siong's constraints also had some names.
- Sorry, the one you mentioned should be another field under Constraint, I forgot to add. The coefficient under ConstraintParameterRelationship is for example "x + 2y <= 0", ConstraintParameter needs the coefficient of 2.
- The reason why we need ConstraintParameterRelationship is because of the coefficient.

File: secretsauce/apps/portal/models.py (hunk 1 above)
Comment (gorold) [reply secretsauce/apps/portal/models.py:36]: ModelTag is for example "Neural Networks" or "Decision Trees", its like a tag to sort different types of Models by.

File: secretsauce/apps/portal/models.py (hunk 2 above)
Comment (gorold) [reply secretsauce/apps/portal/models.py:62]: I was thinking about the case where, what if two DataBlocks share the same schema, and they actually want to reuse the constraints? Will it be better if we link ConstraintBlocks to Projects, then return an error if they choose wrongly?

Comment (eugenechia95) [reply secretsauce/apps/portal/models.py:62]: Okay let's stick to projects then

**Pull request merged**